_includes/tutorials/produce-consume-lang/scala/code/src/main/scala/io/confluent/developer/Consumer.scala

```diff
@@ -0,0 +1,93 @@
+package io.confluent.developer
+
+import io.confluent.developer.Configuration.ConsumerConf
+import io.confluent.developer.schema.ScalaReflectionSerde.reflectionDeserializer4S
+import io.confluent.developer.schema.{Book, ScalaReflectionSerde}
+import io.confluent.kafka.serializers.AbstractKafkaSchemaSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG
+import org.apache.kafka.clients.consumer.{ConsumerRecords, KafkaConsumer}
+import org.apache.kafka.common.serialization.{Deserializer, Serdes}
+import org.apache.kafka.common.utils.Bytes
+import org.slf4j.{Logger, LoggerFactory}
+import pureconfig.ConfigSource
+import pureconfig.generic.auto._
+import ujson.Obj
+
+import scala.collection.mutable
+import scala.concurrent.duration._
+import scala.jdk.CollectionConverters._
+import scala.jdk.DurationConverters._
+
+object Consumer extends cask.MainRoutes with ScalaReflectionSerde with Configuration {
+
+  val bookMap: mutable.Map[String, Book] = mutable.Map[String, Book]()
+
+  private val consumerConf = ConfigSource.default.at("consumer").loadOrThrow[ConsumerConf]
+
+  private val schemaRegistryConfigMap: Map[String, AnyRef] = Map[String, AnyRef](
+    SCHEMA_REGISTRY_URL_CONFIG -> consumerConf.clientConfig.getString(SCHEMA_REGISTRY_URL_CONFIG)
+  )
+
+  override def port: Int = consumerConf.port
+  override def host: String = consumerConf.host
+
+  private def logger: Logger = LoggerFactory.getLogger(getClass)
+
+  @cask.get("/count")
+  def getCount(): Obj = {
+    ujson.Obj("count" -> bookMap.size)
+  }
+
+  @cask.get("/books")
+  def getBooks(): Obj = {
+    ujson.Obj("results" -> ujson.Arr(
+      bookMap.toArray.map { case (_: String, book: Book) =>
+        upickle.default.writeJs(book)
+      }: _*
+    )
+    )
+  }
+
+  def consume(consumer: KafkaConsumer[Bytes, Book]): Vector[Book] = {
+
+    val books: ConsumerRecords[Bytes, Book] = consumer.poll((1 second) toJava)
+
+    books.asScala.toVector.map(_.value())
+  }
+
+  new Thread(() => {
```

Comment:
Hi @bbejeck , Hi @rspurgeon  
I have a suggestion here. The `Consumer#close` inside the shutdown hook has inconsistent behaviours. It randomly raises a `ConcurrentModificationException`.  To solve this we can use a Scheduler. 

```scala
val scheduler: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
  scheduler.schedule(new Runnable {

    override def run(): Unit = {
      logger debug "creating the deserializers and configuring"
      val bookDeserializer: Deserializer[Book] = reflectionDeserializer4S[Book]
      bookDeserializer.configure(schemaRegistryConfigMap.asJava, false)

      logger debug "creating the kafka consumer"
      val consumer = new KafkaConsumer[Bytes, Book](
        consumerConf.clientConfig.toProperties,
        Serdes.Bytes().deserializer(),
        bookDeserializer
      )

      consumer.subscribe(Vector(consumerConf.topics.bookTopic.name).asJava)

      while (!scheduler.isShutdown) {
        Thread.sleep((2 second) toMillis)

        logger debug s"polling the new events"
        val books: Vector[Book] = consume(consumer)

        if (books.nonEmpty) logger info s"just polled ${books.size} books from kafka"
        books.foreach { book =>
          bookMap += book.title -> book
        }
      }

      Try(consumer.close()).recover {
        case error => logger error("Failed to close the kafka consumer", error)
      }
    }
  }, 1, TimeUnit.SECONDS)

  sys.addShutdownHook {
    scheduler.shutdown()
    scheduler.awaitTermination(10, TimeUnit.SECONDS)
  }
```

This will also get ride of the `while(true)`.

Reply:
Feel free to add the change @DivLoic

Reply:
It's done!